src/java.base/share/classes/jdk/internal/foreign/StringSupport.java

```diff
@@ -120,20 +123,47 @@ private static int native_strlen_byte(MemorySegment segment, long start) {
             if (start > 0) {
                 segment = segment.asSlice(start);
             }
-            long len = segment.byteSize() < MAX_TRIVIAL_SIZE
-                    ? (long)STRLEN_TRIVIAL.invokeExact(segment)
-                    : (long)STRLEN.invokeExact(segment);
+            long segmentSize = segment.byteSize();
+            final long len;
+            if (SIZE_T_IS_INT) {
+                if (segmentSize < MAX_TRIVIAL_SIZE) {
+                    len = (int)STRNLEN_TRIVIAL.invokeExact(segment, (int) segmentSize);
+                } else if (segmentSize < Integer.MAX_VALUE) {
+                    len = (int)STRNLEN.invokeExact(segment, (int) segmentSize);
+                } else {
+                    // There is no way to express the max size in the native method using an int so, revert
+                    // to a Java method. It is possible to use a reduction of several STRNLEN invocations
+                    // in a future optimization.
+                    len = strlen_byte(segment);
```

Comment:
Can't we just always throw here? `len` is not allowed to be larger than an `int` any ways.

Reply:
It might be the case that we have a very large segment but there is a zero byte early in the segment.

Reply:
Ok, thanks. That makes sense.